SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorPackageAssembly.cpp

```diff
@@ -0,0 +1,242 @@
+	PackageAssemblyTask = MakeShareable(new FMonitoredProcess(SpatialGDKServicesConstants::SpatialExe, Args, WorkingDir, true));
+	PackageAssemblyTask->OnCompleted().BindSP(this, &FSpatialGDKPackageAssembly::OnTaskCompleted);
+	PackageAssemblyTask->OnOutput().BindSP(this, &FSpatialGDKPackageAssembly::OnTaskOutput);
+	PackageAssemblyTask->OnCanceled().BindSP(this, &FSpatialGDKPackageAssembly::OnTaskCanceled);
+	PackageAssemblyTask->Launch();
+	FString NotificationMessage = FString::Printf(TEXT("Uploading Assembly to Project: %s"), *FSpatialGDKServicesModule::GetProjectName());
+	if (AssemblyDetailsPtr.IsValid())
+	{
+		AssemblyDetailsPtr.Reset();
+	}
+}
+
+void FSpatialGDKPackageAssembly::BuildAllAndUpload(const FString& AssemblyName, const FString& Configuration, const FString& AdditionalArgs, bool bForce)
+{
+	if (AssemblyDetailsPtr == nullptr && Steps.IsEmpty())
+	{
+		AssemblyDetailsPtr.Reset(new AssemblyDetails(AssemblyName, Configuration, bForce));
+		const USpatialGDKEditorSettings* SpatialGDKSettings = GetDefault<USpatialGDKEditorSettings>();
+
+		Steps.Enqueue(EPackageAssemblyStep::BUILD_SERVER);
+		if (SpatialGDKSettings->IsBuildClientWorkerEnabled())
+		{
+			Steps.Enqueue(EPackageAssemblyStep::BUILD_CLIENT);
+		}
+		if (SpatialGDKSettings->IsSimulatedPlayersEnabled())
+		{
+			Steps.Enqueue(EPackageAssemblyStep::BUILD_SIMULATED_PLAYERS);
+		}
+		Steps.Enqueue(EPackageAssemblyStep::UPLOAD_ASSEMBLY);
+
+		AsyncTask(ENamedThreads::GameThread, [this]()
+		{
+			this->ShowTaskStartedNotification(TEXT("Building Assembly"));
+			this->NextStep();
+		});
+	}
+}
+
+bool FSpatialGDKPackageAssembly::CanBuild() const
+{
+	return Steps.IsEmpty();
+}
+
+bool FSpatialGDKPackageAssembly::NextStep()
+{
+	bool HasMoreSteps = false;
+	EPackageAssemblyStep Target = EPackageAssemblyStep::NONE;
+	if (Steps.Dequeue(Target))
+	{
+		HasMoreSteps = true;
+		switch(Target)
+		{
+		case EPackageAssemblyStep::BUILD_SERVER:
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->BuildAssembly(FString::Printf(TEXT("%sServer"), FApp::GetProjectName()), Linux, AssemblyDetailsPtr->Configuration, TEXT(""));
+			});
+			break;
+		case EPackageAssemblyStep::BUILD_CLIENT:
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->BuildAssembly(FApp::GetProjectName(), Win64, AssemblyDetailsPtr->Configuration, TEXT(""));
+			});
+			break;
+		case EPackageAssemblyStep::BUILD_SIMULATED_PLAYERS:
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->BuildAssembly(FString::Printf(TEXT("%sSimulatedPlayer"), FApp::GetProjectName()), Linux, AssemblyDetailsPtr->Configuration, TEXT(""));
+			});
+			break;
+		case EPackageAssemblyStep::UPLOAD_ASSEMBLY:
+		{
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->AssemblyDetailsPtr->Upload(*this);
+			});
+		}
+		break;
+		}
+	}
+	return HasMoreSteps;
+}
+
+void FSpatialGDKPackageAssembly::OnTaskCompleted(int32 TaskResult)
+{
+	if (TaskResult == 0)
```

Comment:
I would prefer to swap this around and early out instead of having multiple if-statements:
```
if (TaskResult != 0)
{
  ..
  // Show that task failed
  return;
}

if (!NextStep())
{
...
}

```

Reply:
Hmm I feel in this case it's clearer to have the two branches for success and failure.